File: python/201810/island/main.cpp
#include <stdio.h>
#include "pi.h"

/*
### with c++ calls only:
FFF='-D_NOFORTRAN'
g++ $FFF -c pi.cpp -o pi.cpp.o
g++ -fPIC -shared -o libpi_cpp.so pi.cpp

g++ $FFF -c main.cpp -o main.cpp.o
g++ -o c.exe main.cpp.o -Wl,-rpath,$PWD/ libpi_cpp.dylib
./c.exe
# pi computed by c = 3.142096
*/

int main()
{
    const int num_points=10000;
    printf("num points=%d\n", num_points);
    printf("pi computed by c = %f\n", approximate_pi_c(num_points));
#ifndef _NOFORTRAN
    printf("pi computed by fortran = %f\n", approximate_pi_fortran(num_points));
#endif

    return 0;
}

File: python/201810/island/main.f90
program main

!### with fortran calls only:
! FFF='-cpp -D_NOC -fno-leading-underscore' # -fno-underscoring
! gfortran $FFF -c pi.f90 -o pi.f90.o
! nm pi.f90.o |grep pi_f
! gfortran -shared -o libpi_fortran.dylib pi.f90.o
! ---
! FFF='-cpp -D_NOC'
! gfortran $FFF -c pi.f90 -o pi.f90.o
! gfortran -shared -o libpi_fortran.dylib pi.f90.o
! gfortran $FFF -c main.f90 -o main.f90.o
! gfortran -o f.exe main.f90.o -Wl,-rpath,$PWD/ libpi_fortran.dylib
! ./f.exe
! # pi computed by fortran =    3.1040000915527344

    use, intrinsic :: iso_c_binding, only: c_double, c_int
    use pi, only: approximate_pi_fortran

    implicit none

#ifndef _NOC
    interface approximate_pi_c
        function approximate_pi_c(num_points) bind (c)
            import :: c_double, c_int
            integer(c_int), value :: num_points
            real(c_double) :: approximate_pi_c
        end function
    end interface
#endif

    integer, parameter :: num_points=10000
    print *,"num points=", num_points 
    print *,"pi computed by fortran = ", approximate_pi_fortran(num_points)
#ifndef _NOC
    print *, "pi computed by c = ", approximate_pi_c(num_points)
#endif

end program
